model/builder/workflow.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2023 The Serverless Workflow Specification Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builder
+
+import (
+	"fmt"
+
+	"github.com/serverlessworkflow/sdk-go/v2/model"
+)
+
+type Builder[T, K any] struct {
+	builder build[K]
+}
+
+func (b *Builder[T, K]) Build() K {
+	return b.builder.Build()
+}
+
+type build[T any] interface {
+	Build() T
+}
+
+// NewWorkflowBuilder create a Workflow Builder
+func NewWorkflowBuilder(id, version string) (*WorkflowBuilder, *Builder[WorkflowBuilder, model.Workflow]) {
+	workflow := model.Workflow{
+		BaseWorkflow: model.BaseWorkflow{
+			ID:          id,
+			SpecVersion: "0.8",
+			Version:     version,
+		},
+	}
+	workflow.ApplyDefault()
+
+	workflowBuilder := &WorkflowBuilder{
+		workflow:              workflow,
+		functions:             make([]*FunctionBuilder[WorkflowBuilder], 0),
+		events:                make([]*EventBuilder[WorkflowBuilder], 0),
+		operatorStateBuilders: make([]*OperationStateBuilder[WorkflowBuilder], 0),
+		eventStateBuilders:    make([]*EventStateBuilder[WorkflowBuilder], 0),
+	}
+
+	builder := &Builder[WorkflowBuilder, model.Workflow]{
+		builder: workflowBuilder,
+	}
+
+	return workflowBuilder, builder
+}
+
+type WorkflowBuilder struct {
```

Comment:
What about an interface instead? Hence, we can model different builders if needed and have a clear API.

Also, can we have a `WorkflowBuilder.BuildAsYaml(path)`? This way we can persist into a file programmatically. wdyt?

Reply:
The builder interface has been decoupled from the `WorkflowBuilder`, so it is possible to have an `ActionBuilder`

It is possible to do this as follows:
```go
type Builder[T, K any] struct {
	builder Build[K]
}

func (b *Builder[T, K]) Build() K {
	return b.builder.Build()
}

func (b *Builder[T, K]) BuildAsYaml() string {
	return b.builder.BuildAsYaml()
}

type Build[T any] interface {
	Build() T
	BuildAsYaml() string
}

```
But i

Reply:
Fair enough!